lib/api/formtools/grid.js

```diff
@@ -0,0 +1,263 @@
+
+    }
+
+    function getToolbarItems (req, res, modelName, cb) {
+
+        if (!modelName || typeof(modelName) !== 'string' || !modelName.length) {
+            throw new Error('modelName is required and must be of type string and cannot be emptied.');
+        }
+
+        var model = linz.get('models')[modelName];
+
+        async.waterfall([
+
+            function (callback) {
+                return getSettings(modelName, callback);
+            }
+
+        ], function (err, settings) {
+
+            if (err) {
+                return cb(err);
+            }
+
+            if (!settings.toolbarItems.length) {
+                return cb(null);
+            }
+
+            async.each(settings.toolbarItems, function (item, itemDone) {
+
+                if (!item.renderer) {
+                    return itemDone(null);
+                }
+
+                // this is a custom action, let's execute the function to get the HTML markup
+                item.renderer(req, res, function (itemErr, html) {
+
+                    if (itemErr) {
+                        return itemDone(itemErr);
+                    }
+
+                    item.html = html;
+
+                    return itemDone(null);
+                });
+
+            }, function (itemErr) {
+                return cb(itemErr, settings.toolbarItems);
+            });
+
+        });
+
+    }
+
+    function getFilters (modelName, cb) {
+
+        if (!modelName || typeof(modelName) !== 'string' || !modelName.length) {
+            throw new Error('modelName is required and must be of type string and cannot be emptied.');
+        }
+
+        var model = linz.get('models')[modelName];
+
+        async.waterfall([
+
+            function (callback) {
+                return getSettings(modelName, callback);
+            }
+
+        ], function (err, settings) {
+
+            if (err) {
+                return cb(err);
+            }
+
+            if (!Object.keys(settings.filters).length) {
+                return cb(null);
+            }
+
+            async.each(Object.keys(settings.filters), function (fieldName, filtersDone) {
+
+                // call the filter renderer and update the content with the result
+                settings.filters[fieldName].filter.renderer(fieldName, function (filterErr, result) {
+
+                    if (filterErr) {
+                        return filtersDone(filterErr);
+                    }
+
+                    settings.filters[fieldName].formControls = result;
+
+                    return filtersDone(null);
+
+                });
+
+            }, function (filterErr) {
+
+                return cb(filterErr, settings.filters);
+
+            });
+
+        });
+
+    }
+
+    function getSearchFilters(fieldNames, data, modelName, cb) {
+
+        if (!fieldNames || !Array.isArray(fieldNames) || !fieldNames.length) {
+            throw new Error('fieldNames is required and must be of type array and cannot be emptied.');
```

Comment:
Should read "fieldNames is required and must be of type Array and cannot be empty."